components/scheduler2/generator/tasks.go

```diff
@@ -0,0 +1,117 @@
+package generator
+
+import (
+	"time"
+
+	enumsspb "go.temporal.io/server/api/enums/v1"
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/components/scheduler2/common"
+	"go.temporal.io/server/service/history/hsm"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type (
+	SleepTask struct {
+		deadline time.Time
+	}
+
+	BufferTask struct{}
+
+	// Fired when generator has no actions to buffer and can sleep.
+	EventSleep struct {
+		Node *hsm.Node
+
+		// Sets the wakeup deadline, after which the state machine will move back to BUFFERING
+		Deadline time.Time
+	}
+
+	// Fired when the generator should buffer more events.
+	EventBuffer struct {
+		Node *hsm.Node
+	}
+)
+
+const (
+	TaskTypeSleep  = "scheduler.generator.Sleep"
+	TaskTypeBuffer = "scheduler.generator.Buffer"
```

Comment:
Wondering why you need two tasks here, why don't you buffer in the sleep task under a write lock? It's less DB writes this way. You should only need the buffer task if you plan to do any IO or other type of side effect.

Reply:
> Wondering why you need two tasks here, why don't you buffer in the sleep task under a write lock? It's less DB writes this way. You should only need the buffer task if you plan to do any IO or other type of side effect.

If that's an option (buffering under a write lock), that works for me, since there shouldn't be any regular contention - posted a question on your other comment, re: `ConflictToken` vs write locking.

Reply:
I was thinking about `WAITING` versus `EXECUTING` in the context of the Backfiller, as well, where a user's backfill request will trigger an immediate state transition and task to start executing. For `Generator`, if we don't have a separate `EXECUTING` task to transition to after a user updates, how would we signal the sleeping `Generator` sub-state machine? If re-entering the `WAITING` state works, then we can probably also get rid of `EXECUTING` on the Backfiller.